demonstrations/tutorial_adjoint_diff.py

```diff
@@ -0,0 +1,457 @@
+#
+#
+# Luckily, we already have a built-in PennyLane function for calculating this.
+
+d_op = qml.operation.operation_derivative(ops[0])
+print(d_op)
+
+##############################################################################
+# Now for calculating the derivative!
+#
+# We loop over the reversed operations, just as before.
+#
+# But if the operation has a parameter, we calculate it's derivative and append it
+# to a list before moving on.
+#
+# Since the ``operation_derivative`` function spits back out a matrix instead of an
+# operation, we have to use ``dev._apply_unitary`` instead to create :math:`|\tilde{b}_i\rangle`.
+
+bra = dev._apply_operation(state, M)
+ket = state
+
+grads = []
+
+for op in reversed(ops):
+    op.inv()
+    ket = dev._apply_operation(ket, op)
+
+    # Calculating the derivative
+    if op.num_params != 0:
+        dU = qml.operation.operation_derivative(op)
+
+        bra_temp = dev._apply_unitary(bra, dU, op.wires)
+
+        dM = 2 * np.real(np.vdot(bra_temp, ket))
+        grads.append(dM)
+
+    bra = dev._apply_operation(bra, op)
+    op.inv()
+
+
+# Finally, reverse the order of the gradients
+# since we calculated them in reverse
+grads = grads[::-1]
+
+print("our calculation: ", grads)
+
+pl_calc = qml.grad(circuit)(x)
+print("Comparison: ", pl_calc)
+
+##############################################################################
+# It matches!!!
+#
+# If you want to use adjoint differentiation without having to code up your own
+# method that can support arbitrary circuits, you can use adjoint diff in PennyLane
+# with ``"default.qubit"``
+
+
+@qml.qnode(dev, diff_method="adjoint")
+def circuit_adjoint(a):
+    qml.RX(a[0], wires=0)
+    qml.CNOT(wires=(0, 1))
+    qml.RY(a[1], wires=1)
+    qml.RZ(a[2], wires=1)
+    return qml.expval(M)
+
+
+qml.grad(circuit_adjoint)(x)
+
+
+##############################################################################
+# Time Scaling
+# ~~~~~~~~~~~~
+#
+# In this section, I've deliberately used the same timing framework as the
+# `backpropagation tutorial <https://pennylane.ai/qml/demos/tutorial_backprop.html#time-comparison>`_ .
+# Though I don't repeat that data here, you can compare the graphs.
+#
+# For this section, we need two additional packages: ``timeit`` and ``pyplot``.
+
+import timeit
+import matplotlib.pyplot as plt
+
+n_wires = 4
+
+dev = qml.device("default.qubit", wires=n_wires)
+
+
+@qml.qnode(dev, diff_method="adjoint")
+def circuit(params):
+    qml.templates.StronglyEntanglingLayers(params, wires=range(n_wires))
+    return qml.expval(qml.PauliZ(0) @ qml.PauliZ(1) @ qml.PauliZ(2) @ qml.PauliZ(3))
+
+
+reps = 2
+num = 3
+
+n_layers = range(1, 21)
+
+t_exec = []
+t_grad = []
+ratio = []
+n_params = []
+
+rng = np.random.default_rng(seed=42)
+
+for i_layers in n_layers:
+
+    # set up the parameters
+    param_shape = qml.templates.StronglyEntanglingLayers.shape(n_wires=4, n_layers=i_layers)
+    params = rng.standard_normal(param_shape)
```

Comment:
```suggestion
    params = np.array(rng.standard_normal(param_shape))
```
This change can be reverted with the new release